File: src/main/scala/utn/frba/iasc/injection/ControllerModule.scala
package utn.frba.iasc.injection

import com.google.inject.name.Named
import com.google.inject.{AbstractModule, Provides, Singleton}
import utn.frba.iasc.controller.{AuctionCreatedController, BuyersController}
import utn.frba.iasc.service.{AuctionService, BidService, UserService}
import utn.frba.iasc.utils.IdGen

object ControllerModule extends AbstractModule {
  @Provides
  @Singleton
  @Named("auctionController")
  def auctionController(
    @Named("auctionService") auctionService: AuctionService,
    @Named("bidService") bidService: BidService,
    @Named("idGen") idGen: IdGen
  ): AuctionCreatedController = new AuctionCreatedController(
    auctionService = auctionService,
    bidService = bidService,
    idGen = idGen
  )

  @Provides
  @Singleton
  @Named("userController")
  def buyersController(
    @Named("userService") userService: UserService,
    @Named("idGen") idGen: IdGen
  ): BuyersController = new BuyersController(
    userService = userService,
    idGen = idGen
  )
}

File: babylon-kt/babylon-in/src/main/kotlin/utn/frba/iasc/babylon/config/ControllerModule.kt
package utn.frba.iasc.babylon.config

import com.fasterxml.jackson.databind.ObjectMapper
import com.google.inject.AbstractModule
import com.google.inject.Provides
import com.google.inject.Singleton
import com.google.inject.name.Named
import utn.frba.iasc.babylon.controller.AuctionController
import utn.frba.iasc.babylon.controller.BuyerController
import utn.frba.iasc.babylon.service.AuctionService
import utn.frba.iasc.babylon.service.BuyerService

object ControllerModule : AbstractModule() {
    @Provides
    @Singleton
    @Named("auctionController")
    fun auctionController(
        @Named("objectMapper") objectMapper: ObjectMapper,
        @Named("auctionService") auctionService: AuctionService
    ) = AuctionController(objectMapper, auctionService)

    @Provides
    @Singleton
    @Named("buyerController")
    fun buyerController(
        @Named("objectMapper") objectMapper: ObjectMapper,
        @Named("buyerService") buyerService: BuyerService
    ) = BuyerController(objectMapper, buyerService)
}